### PR TITLE
Implement/local metrics

### LINKED ISSUE #1833: Local queues prometheus metrics
**What would you like to be added**:

Expose Prometheus metrics for local queues, equivalent to the existing cluster queue metrics, but filtered and labeled by local queues.

Similarly to the visibility API, that serves information about pending workloads in local queues, it would be possible to get metrics like like pending workloads, admitted active workloads, resource usage, etc, for local queues.

If cardinality is a concern, those metrics could be exposed behind a feature flag.

**Why is this needed**:

Metrics about local queues can be useful for the batch users persona, so they can have visibility and historical trends about their workloads.

While some metrics are already available for cluster queues, exposing them to the batch users persona presents the following challenges / limitations:
* Cluster queues metrics are global and cannot be filtered by namespaces / tenants
* Querying "cluster-scoped" metrics in secured Prometheus instances is generally only authorised for cluster admin users that have access to all namespaces / tenants

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [x] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by KPostOffice**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Implementation of LQ metrics KEP

#### Which issue(s) this PR fixes:

Fixes #1833 

#### Special notes for your reviewer:

I'm uncertain if I've updated all the metrics in the right places. I still need to write tests, but I figured I'd open the PR as I have it now in case anything is egregiously off. 

#### Does this PR introduce a user-facing change?

```release-note
Addition of configuration that allows users to get prometheus metrics about LQ info, including the LQ status and the status of pending workloads
```

### CODE REVIEW

File: apis/config/v1beta1/configuration_types.go [hunk 1]
```diff
@@ -146,6 +146,24 @@ type ControllerMetrics struct {
...
+	// LocalQueueMetrics is a configuration that provides enabling LocalQueue metrics and its options.
+	// +optional
+	LocalQueueMetrics *LocalQueueMetrics `json:"localQueueMetrics,omitempty"`
+}
+
+// LocalQueueMetrics defines the configuration options for local queue metrics.
+// If left empty, then metrics will expose for all local queues across namespaces.
+type LocalQueueMetrics struct {
+	// Enable is a knob to allow metrics to be exposed for local queues. Defaults to false.
+	Enable bool `json:"enable,omitempty"`
```
Comment (kannon92) [inline apis/config/v1beta1/configuration_types.go:159]: Could we just take the absence of LocalQueueMetrics to mean that this is not enabled?

File: pkg/controller/core/localqueue_controller.go [hunk 2]
```diff
@@ -300,6 +344,19 @@ func (r *LocalQueueReconciler) SetupWithManager(mgr ctrl.Manager, cfg *config.Co
 		Complete(WithLeadingManager(mgr, r, &kueue.LocalQueue{}, cfg))
 }
 
+func localQueueStatusMetricForReason(reason string, queue *kueue.LocalQueue) {
+	switch reason {
```
Comment (kannon92) [inline pkg/controller/core/localqueue_controller.go:348]: I am very confused on this block of code. Reading the correspondence between the reason and the metric string makes me feel like something is wrong..

If nothing is wrong, can we consider making the metric string match the reason?

**Review by kannon92 — COMMENTED** [review]
There is a lot of code here without any unit tests. I'm not sure what we have existing in this repo for testing around metrics but this code is nontrivial so it may be worth thinking of some test.

**Comment by tenzen-y** [comment]
In favor of #3673 
/close

**Pull request closed without merging**